---
title: "FwsCalc"
author: "Christian Parobek"
date: "04/02/2015"
output: html_document
---

We want to calculate Fws for the _P. falciparum_ samples from our amaAbs data. [Link](http://www.nature.com/nature/journal/v487/n7407/full/nature11174.html) for more details on Fws. This Rmd will calculate Fws from a multiVCF.

This Rmd doc involves multiple steps, including:

1. Subsetting the VCF into exon and non-exon files  
2. Properly format the VCF  
3. Extract the read-depth data using regex  
4. Calculate relevant statistics for exons, nonexons, and exon+nonexon regions

```{r, echo=FALSE, engine='bash'}
## Make the necessary exon and non-exon GFF definition files
grep "exon\|##" PlasmoDB-13.0_Pfalciparum3D7.gff | grep -v "gene" | head -n -1 > PlasmoDB-13.0_Pfalciparum3D7_exons.gff
    # Make an exon-only GFF file (includes header)
    # Grep out any gene entries that slipped in because they had "exon" in their description
    # Remove the last line of the file, since it starts with '##' 

## Use bedtools to subset exonic and non-exonic regions 
bedtools intersect -header -a ../variants/all10_HC.pass.vcf -b PlasmoDB-13.0_Pfalciparum3D7_exons.gff > all10_HC.exons.vcf
    # Get only VCF entries in combined.pass.vcf that are in exons
bedtools intersect -header -v -a ../variants/all10_HC.pass.vcf -b PlasmoDB-13.0_Pfalciparum3D7_exons.gff > all10_HC.nonexons.vcf
    # Get only VCF entries in combined.pass.vcf that are outside exons

## Next, use sed to replace #CHROM -> CHROM
sed 's/#CHROM/CHROM/' ../variants/all10_HC.pass.vcf > all.vcf
sed 's/#CHROM/CHROM/' all10_HC.exons.vcf > exons.vcf
sed 's/#CHROM/CHROM/' all10_HC.nonexons.vcf > nonexons.vcf
```

```{r, echo=FALSE}
## LOAD REGEX LIBRARY
library(stringr)

## READ IN THE Pf AND Pv MULTIVCFs
pfAllVCF <- read.table("all.vcf", comment.char="#", header=TRUE)
pfExVCF <- read.table("exons.vcf", comment.char="#", header=TRUE)
pfNexVCF <- read.table("nonexons.vcf", comment.char="#", header=TRUE)

## DEFINE A FUNCTION THAT WILL CALCULATE Fws FOR EACH DATASET

fwsCalc <- function(dataset) {

  ## REMOVE FIRST NINE COLUMNS FROM THE MULTIVCFs
  data <- dataset[-c(1:9)]
  
  ## EXTRACT RELEVANT READ DEPTH DATA, FIRST MATCH
  refCT <- as.data.frame(sapply(data, function(x) str_extract(x, ":[0123456789]+,")))
      # The numbers pre-comma are ref counts
      # Convert to data frame on the fly
  refCT <- sapply(refCT, function(x) str_extract(x, "[0123456789]+"))
      # Clean out the extra chars, leaving only numbers
  refCT <- apply(refCT, c(1,2), as.numeric)
      # Convert to a numeric matrix
  
  altCT <- as.data.frame(sapply(data, function(x) str_extract(x, ",[0123456789]+:")))
      # The numbers post-comma are alt counts
      # Convert to data frame on the fly
  altCT <- sapply(altCT, function(x) str_extract(x, "[0123456789]+"))
      # Clean out the extra chars, leaving only numbers
  altCT <- apply(altCT, c(1,2), as.numeric)
      # Convert to a numeric matrix
  
  ## CALCULATE qs, ps, and Hs, THE PROPORTIONS OF EACH ALLELE IN THE POPULATION
  ps <- rowSums(refCT)/(rowSums(refCT)+rowSums(altCT))
  qs <- rowSums(altCT)/(rowSums(refCT)+rowSums(altCT))
  Hs <- mean(2*ps*qs)
      # Calculate Hs for each variant and take the mean of all variants
  
  ## CALCULATE qw, pw, and Hw, THE PROPORTIONS OF EACH ALLELE IN EACH INDIVIDUAL
  totCT <- refCT + altCT
      # Make a matrix of total counts
  pw <- matrix(, nrow = length(data[,1]), ncol = length(names(data)))
      # Set up pw matrix
  qw <- matrix(, nrow = length(data[,1]), ncol = length(names(data)))
      # Set up qw matrix
  Hw <- matrix(, nrow = length(data[,1]), ncol = length(names(data)))
      # Set up Hw matrix
  
  for (i in 1:length(names(data))) {
    for (j in 1:length(data[,1])) {
      
      pw[j,i] <- refCT[j,i]/totCT[j,i] # Calculate pw per individual and per allele
      qw[j,i] <- altCT[j,i]/totCT[j,i] # Calculate qw per individual and per allele
      Hw[j,i] <- 2*pw[j,i]*qw[j,i] # Calculate Hw per individual and per allele
      
    }
  }
  
  Hw <- colMeans(Hw)
      # Take the column means of Hw matrix, to get a single Hw score for each sample
  
  ## CALCULATE Fws
  1 - Hw/Hs
  
}

## FINALLY, PLOT THE GRAPHS

pf_all_fws <- fwsCalc(pfAllVCF)
pf_ex_fws <- fwsCalc(pfExVCF)
pf_nex_fws <- fwsCalc(pfNexVCF)

hist(pf_all_fws, breaks=14)
hist(pf_ex_fws, breaks=14)
hist(pf_nex_fws, breaks=14)

```

